modules/dnn/src/webgpu/src/buffer.cpp

```diff
@@ -0,0 +1,122 @@
+#include <string.h>
+#include "../dawn/dawnUtils.hpp"
+#include "../../precomp.hpp"
+#include "common.hpp"
+#include "internal.hpp"
+#include "../include/buffer.hpp"
+
+namespace cv { namespace dnn { namespace webgpu {
+#ifdef HAVE_WEBGPU
+
+#ifdef __EMSCRIPTEN__
+    EM_JS(const void*, readBufferAsync, (WGPUBuffer id), { 
+        return Asyncify.handleAsync(async () => {
+            const gpuReadBuffer = WebGPU.mgrBuffer.get(id);
+            await gpuReadBuffer.mapAsync(GPUMapMode.READ);
+            const arrayBuffer = gpuReadBuffer.getMappedRange();
+            console.log(arrayBuffer.length);
+            return arrayBuffer;
```

Comment:
The `ArrayBuffer` doesn't convert to `void*`. You may need to alloc the memory on Wasm heap and copy the data of `arrayBuffer` to that memory. After use, you need to free that memory in your C code.

Here are some references:
https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html#access-memory-from-javascript
https://emscripten.org/docs/api_reference/emscripten.h.html#c.EM_JS